frame/contracts/src/wasm/mod.rs

```diff
@@ -235,77 +233,245 @@ impl<T: Config> PrefabWasmModule<T> {
+	fn increment_refcount(code_hash: CodeHash<T>) -> Result<(), DispatchError> {
+		<CodeInfoOf<T>>::mutate(code_hash, |existing| -> Result<(), DispatchError> {
+			if let Some(info) = existing {
+				info.refcount = info.refcount.saturating_add(1);
+				Ok(())
+			} else {
+				Err(Error::<T>::CodeNotFound.into())
+			}
+		})
+	}
 	/// See [`Self::from_code_unchecked`].
 	#[cfg(feature = "runtime-benchmarks")]
 	pub fn store_code_unchecked(
-		original_code: Vec<u8>,
+		code: Vec<u8>,
 		schedule: &Schedule<T>,
 		owner: T::AccountId,
 	) -> DispatchResult {
-		let executable = Self::from_code_unchecked(original_code, schedule, owner)?;
-		code_cache::store(executable, false)
+		let executable = Self::from_code_unchecked(code, schedule, owner)?;
+		Self::store_code(executable, false)
 	}
 
-	/// Decrement instruction_weights_version by 1. Panics if it is already 0.
-	#[cfg(test)]
-	pub fn decrement_version(&mut self) {
-		self.instruction_weights_version = self.instruction_weights_version.checked_sub(1).unwrap();
-	}
-
-	/// Create the module without checking nor instrumenting the passed code.
+	/// Create the module without checking the passed code.
 	///
 	/// # Note
 	///
-	/// This is useful for benchmarking where we don't want instrumentation to skew
+	/// This is useful for benchmarking where we don't want validation of the module to skew
 	/// our results. This also does not collect any deposit from the `owner`. Also useful
 	/// during testing when we want to deploy codes that do not pass the instantiation checks.
 	#[cfg(any(test, feature = "runtime-benchmarks"))]
 	fn from_code_unchecked(
-		original_code: Vec<u8>,
+		code: Vec<u8>,
 		schedule: &Schedule<T>,
 		owner: T::AccountId,
 	) -> Result<Self, DispatchError> {
-		prepare::benchmarking::prepare(original_code, schedule, owner)
-			.map_err::<DispatchError, _>(Into::into)
+		prepare::benchmarking::prepare(code, schedule, owner)
 	}
 }
 
-impl<T: Config> OwnerInfo<T> {
+impl<T: Config> CodeInfo<T> {
 	/// Return the refcount of the module.
 	#[cfg(test)]
 	pub fn refcount(&self) -> u64 {
 		self.refcount
 	}
 }
 
-impl<T: Config> Executable<T> for PrefabWasmModule<T> {
+impl<T: Config> Executable<T> for WasmBlob<T> {
 	fn from_storage(
 		code_hash: CodeHash<T>,
-		schedule: &Schedule<T>,
 		gas_meter: &mut GasMeter<T>,
 	) -> Result<Self, DispatchError> {
-		code_cache::load(code_hash, schedule, gas_meter)
+		let code = Self::load_code(code_hash, gas_meter)?;
+		// We store `code_info` at the same time as contract code,
+		// therefore this query shouldn't really fail.
+		// We consider its failure equal to `CodeNotFound`, as contract code without
+		// `code_info` is unusable in this pallet.
+		let code_info = <CodeInfoOf<T>>::get(code_hash).ok_or(Error::<T>::CodeNotFound)?;
+
+		Ok(Self { code, code_info, code_hash })
 	}
 
 	fn add_user(code_hash: CodeHash<T>) -> Result<(), DispatchError> {
-		code_cache::increment_refcount::<T>(code_hash)
+		Self::increment_refcount(code_hash)
```

Comment:
why do we need both add_user and increment_refcount and their counter parts (remove_user / decrement_refcount) can't we just keep increment decrement_refcount and delete the others ?

Reply:
Yes sure merge them. We no longer need both as it seems.

Reply:
merged